File: Task/The-ISAAC-cipher/Fortran/the-isaac-cipher.f
module isaac_cipher
    implicit none
    integer, parameter :: iEncrypt = 0, iDecrypt = 1
    integer, parameter :: n = 256
    character(len=*), parameter :: msg = 'a Top Secret secret'
    character(len=*), parameter :: key = 'this is my secret key'

    integer :: randcnt
    integer :: aa, bb, cc
    integer, dimension(0:n-1) :: randrsl, mm

contains

subroutine isaac()
    integer :: i, x, y
    cc = cc + 1
    bb = bb + cc

    do i = 0, n-1
        x = mm(i)
        select case(mod(i, 4))
        case(0)
            aa = ieor(aa, ishft(aa, 13))
        case(1)
            aa = ieor(aa, ishft(aa, -6))
        case(2)
            aa = ieor(aa, ishft(aa, 2))
        case(3)
            aa = ieor(aa, ishft(aa, -16))
        end select
        aa = mm(iand(i + 128, 255)) + aa
        y = mm(iand(ishft(x, -2), 255)) + aa + bb
        mm(i) = y
        bb = mm(iand(ishft(y, -10), 255)) + x
        randrsl(i) = bb
    end do
    randcnt = 0
end subroutine isaac

subroutine mix(a, b, c, d, e, f, g, h)
    integer, intent(inout) :: a, b, c, d, e, f, g, h
    a = ieor(a, ishft(b, 11)); d = d + a; b = b + c
    b = ieor(b, ishft(c, -2)); e = e + b; c = c + d
    c = ieor(c, ishft(d, 8)); f = f + c; d = d + e
    d = ieor(d, ishft(e, -16)); g = g + d; e = e + f
    e = ieor(e, ishft(f, 10)); h = h + e; f = f + g
    f = ieor(f, ishft(g, -4)); a = a + f; g = g + h
    g = ieor(g, ishft(h, 8)); b = b + g; h = h + a
    h = ieor(h, ishft(a, -9)); c = c + h; a = a + b
end subroutine mix

subroutine irandinit(flag)
    logical, intent(in) :: flag
    integer :: i, a, b, c, d, e, f, g, h
    aa = 0; bb = 0; cc = 0
    a = int(z'9e3779b9')
    b = a; c = a; d = a; e = a; f = a; g = a; h = a

    do i = 1, 4
        call mix(a, b, c, d, e, f, g, h)
    end do

    i = 0
    do while(i < n)
        if (flag) then
            a = a + randrsl(i);   b = b + randrsl(i+1)
            c = c + randrsl(i+2); d = d + randrsl(i+3)
            e = e + randrsl(i+4); f = f + randrsl(i+5)
            g = g + randrsl(i+6); h = h + randrsl(i+7)
        endif
        call mix(a, b, c, d, e, f, g, h)
        mm(i:i+7) = [a, b, c, d, e, f, g, h]
        i = i + 8
    end do

    if (flag) then
        i = 0
        do while(i < n)
            a = a + mm(i);   b = b + mm(i+1)
            c = c + mm(i+2); d = d + mm(i+3)
            e = e + mm(i+4); f = f + mm(i+5)
            g = g + mm(i+6); h = h + mm(i+7)
            call mix(a, b, c, d, e, f, g, h)
            mm(i:i+7) = [a, b, c, d, e, f, g, h]
            i = i + 8
        end do
    endif
    call isaac()
end subroutine irandinit

subroutine iseed(seed, flag)
    character(len=*), intent(in) :: seed
    logical, intent(in) :: flag
    integer :: i, m

    mm = 0
    m = len(seed)
    do i = 0, n-1
        if (i < m) then
            randrsl(i) = ichar(seed(i+1:i+1))
        else
            randrsl(i) = 0
        endif
    end do
    call irandinit(flag)
end subroutine iseed

function irandom() result(res)
    integer :: res
    res = randrsl(randcnt)
    randcnt = randcnt + 1
    if (randcnt >= n) then
        call isaac()
        randcnt = 0
    endif
end function irandom

function iranda() result(res)
    integer :: res
    res = mod(irandom(), 95) + 32
end function iranda

function ascii2hex(s) result(res)
    character(len=*), intent(in) :: s
    character(len=2*len(s)) :: res
    integer :: i, val

    res = ''
    do i = 1, len(s)
        val = ichar(s(i:i))
        write(res(2*i-1:2*i), '(Z2.2)') val
    end do
end function ascii2hex

function vernam(msg) result(res)
    character(len=*), intent(in) :: msg
    character(len=len(msg)) :: res
    integer :: i, c

    res = ''
    do i = 1, len(msg)
        c = ieor(iranda(), ichar(msg(i:i)))
        res(i:i) = char(c)
    end do
end function vernam

function caesar(m, ch, shift, modulo, start) result(res)
    integer, intent(in) :: m, shift, modulo
    character, intent(in) :: ch, start
    character :: res
    integer :: n, s

    s = shift
    if (m == iDecrypt) s = -s
    n = mod(ichar(ch) - ichar(start) + s, modulo)
    if (n < 0) n = n + modulo
    res = char(ichar(start) + n)
end function caesar

function vigenere(msg, m) result(res)
    character(len=*), intent(in) :: msg
    integer, intent(in) :: m
    character(len=len(msg)) :: res
    integer :: i

    res = ''
    do i = 1, len(msg)
        res(i:i) = caesar(m, msg(i:i), iranda(), 95, ' ')
    end do
end function vigenere

end module isaac_cipher

program rosetta_isaac
    use isaac_cipher
    implicit none
    character(len=len(msg)) :: xctx, mctx, xptx, mptx

    call iseed(key, .true.)
    xctx = vernam(msg)
    mctx = vigenere(msg, iEncrypt)

    call iseed(key, .true.)
    xptx = vernam(xctx)
    mptx = vigenere(mctx, iDecrypt)

    print '(A)', 'Message: '//msg
    print '(A)', 'Key    : '//key
    print '(A)', 'XOR    : '//ascii2hex(xctx)
    print '(A)', 'MOD    : '//ascii2hex(mctx)
    print '(A)', 'XOR dcr: '//xptx
    print '(A)', 'MOD dcr: '//mptx
end program rosetta_isaac

File: Task/The-ISAAC-cipher/Nim/the-isaac-cipher.nim
import strutils

type

  IMode = enum iEncrypt, iDecrypt

  State = object
    # Internal.
    mm: array[256, uint32]
    aa, bb, cc: uint32
    # External.
    randrsl: array[256, uint32]
    randcnt: uint32


proc isaac(s: var State) =

  inc s.cc        # "cc" just gets incremented once per 256 results
  s.bb += s.cc    # then combined with "bb".

  for i in 0u32..255:
    let x = s.mm[i]
    case range[0..3](i and 3)
    of 0: s.aa = s.aa xor s.aa shl 13
    of 1: s.aa = s.aa xor s.aa shr 6
    of 2: s.aa = s.aa xor s.aa shl 2
    of 3: s.aa = s.aa xor s.aa shr 16
    s.aa += s.mm[(i + 128) and 255]
    let y = s.mm[(x shr 2) and 255] + s.aa + s.bb
    s.mm[i] = y
    s.bb = s.mm[(y shr 10) and 255] + x
    s.randrsl[i] = s.bb

  s.randcnt = 0


proc mix(a: var array[8, uint32]) =
  a[0] = a[0] xor a[1] shl 11; a[3] += a[0]; a[1] += a[2]
  a[1] = a[1] xor a[2] shr  2; a[4] += a[1]; a[2] += a[3]
  a[2] = a[2] xor a[3] shl  8; a[5] += a[2]; a[3] += a[4]
  a[3] = a[3] xor a[4] shr 16; a[6] += a[3]; a[4] += a[5]
  a[4] = a[4] xor a[5] shl 10; a[7] += a[4]; a[5] += a[6]
  a[5] = a[5] xor a[6] shr  4; a[0] += a[5]; a[6] += a[7]
  a[6] = a[6] xor a[7] shl  8; a[1] += a[6]; a[7] += a[0]
  a[7] = a[7] xor a[0] shr  9; a[2] += a[7]; a[0] += a[1]


proc iRandInit(s: var State; flag: bool) =

  s.aa = 0; s.bb = 0; s.cc = 0
  var a: array[8, uint32]
  for item in a.mitems: item = 0x9e3779b9u32  # The golden ratio.

  for i in 0..3:  # Scramble it.
    a.mix()

  var i = 0u32
  while true:   # Fill in "mm" with messy stuff.
    if flag:
      # Use all the information in the seed.
      for n in 0u32..7: a[n] += s.randrsl[n + i]
    a.mix()
    for n in 0u32..7: s.mm[n + i] = a[n]
    inc i, 8
    if i > 255: break

  if flag:
    # Do a second pass to make all of the seed affect all of "mm".
    i = 0
    while true:
      for n in 0u32..7: a[n] += s.mm[n + i]
      a.mix()
      for n in 0u32..7: s.mm[n + i] = a[n]
      inc i, 8
      if i > 255: break

  s.isaac()       # Fill in the first set of results.
  s.randcnt = 0   # Prepare to use the first set of results.


proc iSeed(s: var State; seed: string; flag: bool) =
  ## Seed ISAAC with a given string.
  ## The string can be any size. The first 256 values will be used.
  s.mm.reset()
  let m = seed.high
  for i in 0..255:
    s.randrsl[i] = if i > m: 0 else: ord(seed[i])
  # Initialize ISAAC with seed.
  s.iRandInit(flag)


proc iRandom(s: var State): uint32 =
  ## Get a random 32-bit value 0..int32.high.
  result = s.randrsl[s.randcnt]
  inc s.randcnt
  if s.randcnt > 255:
    s.isaac()
    s.randcnt = 0


proc iRandA(s: var State): byte =
  ## Get a random character in printable ASCII range.
  result = byte(s.iRandom() mod 95 + 32)


proc vernam(s: var State; msg: string): string =
  ## XOR encrypt on random stream. Output: ASCII string.
  result.setLen(msg.len)
  for i, c in msg:
    result[i] = chr(s.irandA() xor byte(c))


template letterNum(letter, start: char): int =
  ord(letter) - ord(start)


proc caesar(m: IMode; ch: char; shift, modulo: int; start: char): char =
  let shift = if m == iEncrypt: shift else: -shift
  var n = letterNum(ch, start) + shift
  n = n mod modulo
  if n < 0: inc n, modulo
  result = chr(ord(start) + n)


proc vigenere(s: var State; msg: string; m: IMode): string =
  ## Vigenere MOD 95 encryption & decryption. Output: ASCII string.
  result.setLen(msg.len)
  for i, c in msg:
    result[i] = caesar(m, c, s.iRanda().int, 95, ' ')


let
  msg = "a Top Secret secret"
  key = "this is my secret key"

var state: State

# 1) seed ISAAC with the key
state.iSeed(key, true)
# 2) Encryption
# a) XOR (Vernam)
let xctx = state.vernam(msg)                # XOR ciphertext.
# b) MOD (Vigenere)
let mctx = state.vigenere(msg, iEncrypt)    # MOD ciphertext.
# 3) Decryption
state.iSeed(key, true)
# a) XOR (Vernam)
let xptx = state.vernam(xctx)               # XOR decryption (plaintext).
# b) MOD (Vigenere)
let mptx = state.vigenere(mctx, iDecrypt)   # MOD decryption (plaintext).
# Program output
echo "Message: ", msg
echo "    Key: ", key
echo "    XOR: ", xctx.tohex
echo "    MOD: ", mctx.toHex
echo "XOR dcr: ", xptx
echo "MOD dcr: ", mptx
